process/block/baseProcess.go

```diff
@@ -2585,23 +2585,37 @@ func (bp *baseProcessor) OnProposedBlock(
 		return process.ErrWrongTypeAssertion
 	}
 
-	// TODO: call SetRootHashIfNeeded for accountsProposal which should recreate the trie if needed for
-	accountsProvider, err := state.NewAccountsEphemeralProvider(bp.accountsProposal)
+	lastExecResHandler, err := bp.prepareAccountsForProposal()
 	if err != nil {
 		return err
 	}
 
-	lastCommittedHeader, err := bp.dataPool.Headers().GetHeaderByHash(proposedHeader.GetPrevHash())
+	accountsProvider, err := state.NewAccountsEphemeralProvider(bp.accountsProposal)
 	if err != nil {
 		return err
 	}
 
-	lastExecResHandler, err := common.GetOrCreateLastExecutionResultForPrevHeader(lastCommittedHeader, proposedHeader.GetPrevHash())
+	return bp.dataPool.Transactions().OnProposedBlock(proposedHash, proposedBodyPtr, proposedHeader, accountsProvider, lastExecResHandler.GetHeaderHash())
+}
+
+func (bp *baseProcessor) prepareAccountsForProposal() (data.BaseExecutionResultHandler, error) {
+	prevHeader := bp.blockChain.GetCurrentBlockHeader()
+	prevHeaderHash := bp.blockChain.GetCurrentBlockHeaderHash()
+	if check.IfNil(prevHeader) || len(prevHeaderHash) == 0 {
```

Comment:
Perhaps it would be clearer to return `ErrNilHash` if `len(prevHeaderHash) == 0` instead of `ErrNilHeaderHandler`